### ISSUE TITLE
DemoteF64ToF32Pass corrupts signature of imported functions

### ISSUE THREAD

**Issue by ezhulenev** [iree-org/iree#12987]
### What happened?

```mlir
  func.func private @cudnn.pointwise_relu(
    %input: !my.type, %lower: f64, %upper: f64
  ) -> !my.type
```

gets converted to 

```mlir
  func.func private @cudnn.pointwise_relu(
    %input: !my.type, %lower: f32, %upper: f32
  ) -> !my.type
```

and then fails calling convention validation at model loading time.

All type conversion passes should take extra care of imported functions, however if imported function has `tensor<f64>` argument, presumably it has to be demoted to `tensor<f32>` together with every other tensor in the program?

**Comment by stellaraccident** [comment]
Ugh. I hate that pass. Thank you for the catch.

**Comment by benvanik** [comment]
We should just bail on any external function - we can't change their types as we don't control the implementation.

**Comment by allieculp** [comment]
From triage meeting: setting as P1 - needs scheduling. 
@ezhulenev Are you able to pick this up?

**Comment by ezhulenev** [comment]
I'll try to get to it maybe next week, it's not blocking anything for me (I just use f32 everywhere for now)

**Comment by allieculp** [comment]
Sounds good, thanks! Please update the status when you get to it.

**Comment by allieculp** [comment]
@ezhulenev Curious if there is any news on this one?

**Comment by ezhulenev** [comment]
No, not blocking any other work, plan to get back to it only once will hit this problem once again as a part of cudnn work

**Comment by ezhulenev** [comment]
@bviyer this issue came up in cudnn integration, we'll need to change f32 to f64 here: https://github.com/openxla/openxla-nvgpu/blob/628fc75bcfe00c0ab2551824914a896668f21909/compiler/src/openxla/compiler/nvgpu/Dialect/CUBLAS/IR/CUBLASOps.td#L74-L75, and make sure it works end-to-end, this also will be needed for cublas integration as we'll have to pass f64 alpha/beta as well

**Comment by bviyer** [comment]
@ezhulenev Other than the test in the "What Happened?" section, do you have a larger test that is failing that I can make sure will pass with a fix?

**Comment by ezhulenev** [comment]
No, I worked around this issue in cudnn plugin by using F32 type, success criteria should be "cudnn uses F64 for alpha and beta and everything works", just try to change it to F64 and everything will stop working :)
